pkg/odo/cli/utils/migrate.go

```diff
@@ -0,0 +1,349 @@
+)
+
+// TODO
+// lables and annotations that are added on s2i components
+
+const (
+	migrateCommandName = "migrate-to-devfile"
+	//build command id to be used in s2i devfile
+	buildCommandID = "s2i-assemble"
+	// build command to be used in s2i devfile
+	buildCommandS2i = "/opt/odo/bin/s2i-setup && /opt/odo/bin/assemble-and-restart"
+	// run command id to be used in devfile
+	runCommandID = "s2i-run"
+	// run command to be used in s2i devfile
+	runCommandS2i = "/opt/odo/bin/run"
+	// container component name to be used in devfile
+	containerName = "s2i-builder"
+	// directory to sync s2i source code
+	sourceMappingS2i = "/tmp/projects"
+	// devfile version
+	devfileVersion = "2.0.0"
+	// environment variable set for s2i assemble and restart scripts
+	// some change in script if scripts is executed for a s2i component converted to devfile
+	envS2iConvertedDevfile = "ODO_S2I_CONVERTED_DEVFILE"
+)
+
+var migrateLongDesc = ktemplates.LongDesc(`Migrate S2I components to devfile components`)
+
+var migrateExample = ktemplates.Examples(`odo utils migrate-to-devfile`)
+
+// MigrateOptions encapsulates the options for the command
+type MigrateOptions struct {
+	context          *genericclioptions.Context
+	componentContext string
+	componentName    string
+}
+
+// NewMigrateOptions creates a new MigrateOptions instance
+func NewMigrateOptions() *MigrateOptions {
+	return &MigrateOptions{}
+}
+
+// Complete completes MigrateOptions after they've been created
+func (mo *MigrateOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+
+	if !util.CheckPathExists(component.ConfigFilePath) {
+		return errors.New("this directory does not contain an odo s2i component, Please run the command from odo component directory to migrate s2i component to devfile")
+	}
+
+	mo.context = genericclioptions.NewContext(cmd)
+	mo.componentContext = component.LocalDirectoryDefaultLocation
+	mo.componentName = mo.context.LocalConfigInfo.GetName()
+	return nil
+
+}
+
+// Validate validates the MigrateOptions based on completed values
+func (mo *MigrateOptions) Validate() (err error) {
+	if mo.context.LocalConfigInfo.GetSourceType() == config.GIT {
+		return errors.New("migration of git type s2i components to devfile not supported by odo")
+	}
+
+	return nil
+}
+
+// Run contains the logic for the command
+func (mo *MigrateOptions) Run() (err error) {
+
+	/*  This data is yet to be converted
+
+	// Absolute path
+	sourcePath, _ := context.LocalConfigInfo.GetOSSourcePath()
+	minMemory := context.LocalConfigInfo.GetMinMemory()
+	minCPU := context.LocalConfigInfo.GetMinCPU()
+	maxCPU := context.LocalConfigInfo.GetMaxCPU()
+
+	*/
+
+	err = generateDevfileYaml(mo)
+	if err != nil {
+		return errors.Wrap(err, "Error in generating devfile.yaml")
+	}
+
+	err = generateEnvYaml(mo)
+	if err != nil {
+		return errors.Wrap(err, "Error in generating env.yaml")
+	}
+
+	// TODO: Delete the s2i component and deploy the devfile component.
```

Comment:
TODO found

Reply:
It is just a note about the data we are unable to migrate.

Reply:
Please update with some reasoning.